crates/flashblocks/builder/src/validator.rs

```diff
@@ -0,0 +1,776 @@
+                target: "flashblocks::state_executor",
+                ?access_list_hash,
+                expected = ?access_list_hash,
+                access_list = ?computed_access_list,
+                expected_access_list = ?access_list.clone(),
+                "Access list hash mismatch"
+            );
+
+            return Err(BalExecutorError::BalHashMismatch {
+                expected: access_list_hash,
+                got: computed_access_list_hash,
+            });
+        }
+
+        if outcome.block.receipts_root != diff.receipts_root {
+            error!(
+                target: "flashblocks::state_executor",
+                got = ?outcome.block.receipts_root,
+                expected = ?diff.receipts_root,
+                "Receipts root mismatch"
+            );
+
+            return Err(BalExecutorError::ReceiptsRootMismatch {
+                expected: diff.receipts_root,
+                got: outcome.block.receipts_root,
+            });
+        }
+
+        if outcome.block.state_root != diff.state_root {
+            error!(
+                target: "flashblocks::state_executor",
+                got = ?outcome.block.state_root,
+                expected = ?diff.state_root,
+                "State root mismatch"
+            );
+
+            return Err(BalExecutorError::StateRootMismatch {
+                expected: diff.state_root,
+                got: outcome.block.state_root,
+            });
+        }
+
+        if outcome.block.hash() != diff.block_hash {
+            error!(
+                target: "flashblocks::state_executor",
+                got = ?outcome.block.hash(),
+                expected = ?diff.block_hash,
+                "Block hash mismatch"
+            );
+
+            return Err(BalExecutorError::BalHashMismatch {
+                expected: diff.block_hash,
+                got: outcome.block.hash(),
+            });
+        }
+
+        // 6. Seal the block
+        let BlockBuilderOutcome {
+            execution_result,
+            block,
+            hashed_state,
+            trie_updates,
+        } = outcome;
+
+        let sealed_block = Arc::new(block.sealed_block().clone());
+
+        let execution_outcome = ExecutionOutcome::new(
+            state.take_bundle(),
+            vec![execution_result.receipts.clone()],
+            block.number(),
+            Vec::new(),
+        );
+
+        let executed_block: BuiltPayloadExecutedBlock<OpPrimitives> = BuiltPayloadExecutedBlock {
+            recovered_block: Arc::new(block),
+            execution_output: Arc::new(execution_outcome),
+            hashed_state: either::Left(Arc::new(hashed_state)),
+            trie_updates: either::Left(Arc::new(trie_updates)),
+        };
+
+        let payload = OpBuiltPayload::new(
+            payload_id,
+            sealed_block,
+            U256::from(fees),
+            Some(executed_block),
+        );
+
+        Ok(payload)
+    }
+
+    fn validate_flashblock_sequential(
```

Comment:
For now, we should probably do the bal verification and the non-bal verification in parallel. If the bal verification fails then we can fallback to the non bal.